packages/core/src/render3/hooks.ts

```diff
@@ -103,50 +110,74 @@ export function registerPostOrderHooks(tView: TView, tNode: TNode): void {
+ * some recursively trigger a change detection cycle.
+ * To solve that, it is required to track the state of the execution of these init hooks.
+ * This is done by storing and maintaining flags in the view: the {@link InitPhaseState},
+ * and the index within that phase. They can be seen as a cursor in the following structure:
+ * [[onInit1, onInit2], [afterContentInit1], [afeterViewInit1, afterViewInit2, afterViewInit3]]
+ * They are are stored as flags in LView[FLAGS].
+ *
+ * 2. Pre-order hooks can be executed in batches, because of the flushHooksUpTo instruction.
+ * To be able to pause and resume their execution, we also need some state about the hook's array
+ * that is being processed:
+ * - the index of the next hook to be executed
+ * - the number of init hooks already found in the processed part of the  array
+ * They are are stored as flags in LView[FLAGS_MORE].
+ */
+
 /**
  * Executes necessary hooks at the start of executing a template.
  *
  * Executes hooks that are to be run during the initialization of a directive such
  * as `onChanges`, `onInit`, and `doCheck`.
  *
- * Has the side effect of updating the RunInit flag in `lView` to be `0`, so that
- * this isn't run a second time.
- *
  * @param lView The current view
  * @param tView Static data for the view containing the hooks to be executed
  * @param checkNoChangesMode Whether or not we're in checkNoChanges mode.
+ * @param currentNodeIndex see below
  */
-export function executeInitHooks(
-    currentView: LView, tView: TView, checkNoChangesMode: boolean): void {
+export function executePreOrderHooks(
+    currentView: LView, tView: TView, checkNoChangesMode: boolean,
+    currentNodeIndex?: number): void {
   if (!checkNoChangesMode) {
     executeHooks(
-        currentView, tView.initHooks, tView.checkHooks, checkNoChangesMode,
-        InitPhaseState.OnInitHooksToBeRun);
+        currentView, tView.preOrderHooks, tView.preOrderCheckHooks, checkNoChangesMode,
+        InitPhaseState.OnInitHooksToBeRun,
+        currentNodeIndex !== undefined ? currentNodeIndex : null);
   }
 }
 
 /**
  * Executes hooks against the given `LView` based off of whether or not
  * This is the first pass.
  *
- * @param lView The view instance data to run the hooks against
+ * @param currentView The view instance data to run the hooks against
  * @param firstPassHooks An array of hooks to run if we're in the first view pass
  * @param checkHooks An Array of hooks to run if we're not in the first view pass.
  * @param checkNoChangesMode Whether or not we're in no changes mode.
+ * @param initPhaseState the current state of the init phase
+ * @param currentNodeIndex see below
  */
 export function executeHooks(
     currentView: LView, firstPassHooks: HookData | null, checkHooks: HookData | null,
-    checkNoChangesMode: boolean, initPhase: number): void {
+    checkNoChangesMode: boolean, initPhaseState: InitPhaseState,
+    currentNodeIndex?: number | null): void {
```

Comment:
Let's avoid adding optional args because they create wrapper functions

```suggestion
    currentNodeIndex: number | null): void {
```

Reply:
Unless I'm missing something, they don't.
I think it is cleaner to keep this parameter optional.

Reply:
This is something Chrome does internally (not Typescript). This is something Misko told me recently, so maybe he can provide more details (cc @mhevery)

Reply:
I see, I was missing something! I'll update that.